Add --add-opens java.base/java.lang=ALL-UNNAMED to MiniMix_5min for 11+

Mesbah-Alam opened the pull request:
Resolves https://github.com/eclipse/openj9/issues/11607

Signed-off-by: Mesbah_Alam@ca.ibm.com <Mesbah_Alam@ca.ibm.com>

Mesbah-Alam commented:
Test with JDK16 : Grinder/12985
Test with JDK8: Grinder/12984

Mesbah-Alam commented:
This PR should resolve these failures too: https://trss.adoptopenjdk.net/output/test?id=5ff71a6b2bad78506603a076

llxia commented on system/otherLoadTest/playlist.xml:
```diff
@@ -22,6 +22,23 @@
 		<variations>
 			<variation>NoOptions</variation>
 		</variations>
+		<command>$(SYSTEMTEST_CMD_TEMPLATE) -test=MixedLoadTest -test-args=$(Q)timeLimit=5m$(Q) -java-args=$(Q)--add-opens java.base/java.lang=ALL-UNNAMED$(Q);\
```
` -java-args` is already set in `$(SYSTEMTEST_CMD_TEMPLATE)`. In the console output, `-java-args` set twice 
- `-java-args=`
- `-java-args="--add-opens java.base/java.lang=ALL-UNNAMED"`

This worked in Grinder because we are running no options. Once we enable other JVM options, I suspect this will break. Unless STF aggregates all `-java-args` values.

```
00:04:10.165  mkdir -p "/home/jenkins/workspace/Grinder/openjdk-tests/TKG/../TKG/test_output_16101271418479/MiniMix_5min_0"; \
00:04:10.166  cd "/home/jenkins/workspace/Grinder/openjdk-tests/TKG/../TKG/test_output_16101271418479/MiniMix_5min_0"; \
00:04:10.166  perl /home/jenkins/workspace/Grinder/openjdk-tests/TKG/../../jvmtest/system/otherLoadTest/..//stf/stf.core/scripts/stf.pl -test-root="/home/jenkins/workspace/Grinder/openjdk-tests/TKG/../../jvmtest/system/otherLoadTest/..//stf;/home/jenkins/workspace/Grinder/openjdk-tests/TKG/../../jvmtest/system/otherLoadTest/..//openjdk-systemtest;/home/jenkins/workspace/Grinder/openjdk-tests/TKG/../../jvmtest/system/otherLoadTest/..//openj9-systemtest" -systemtest-prereqs="/home/jenkins/workspace/Grinder/openjdk-tests/TKG/../../jvmtest/system/otherLoadTest/..//systemtest_prereqs" -java-args=' ' -results-root="/home/jenkins/workspace/Grinder/openjdk-tests/TKG/../TKG/test_output_16101271418479/MiniMix_5min_0" -test=MixedLoadTest -test-args="timeLimit=5m" -java-args="--add-opens java.base/java.lang=ALL-UNNAMED";\

```

Mesbah-Alam replied:
Great catch @llxia . I've updated the PR to use a different STF option (-java-args-execute-initial) to supply the additional JVM option. 

Test : Grinder/12991 

So, now we have both the options coming in from Mode150 (e.g. `-XX:+UseCompressedOops`), as well as `--add-opens java.base/java.lang=ALL-UNNAMED`

```
14:10:02  STF 11:10:01.596 - Running command: /home/jenkins/workspace/Grinder/openjdkbinary/j2sdk-image/bin/java --add-opens java.base/java.lang=ALL-UNNAMED -XX:+UseCompressedOops
```

llxia commented:
https://github.com/AdoptOpenJDK/openjdk-tests/pull/2148 is merged. Please rebase your branch. MiniMix_5min runs two variations now.

Mesbah-Alam commented:
@pshipton :  

With `--add-opens java.base/java.lang=ALL-UNNAMED` added, MiniMix test starts to work fine on openJ9 JDK16 (e.g. internal Grinder/12991). 

However, it [fails on hotSpot JDK16](https://ci.adoptopenjdk.net/view/Test_grinder/job/Grinder/5882/consoleFull ) with following error still: 

```
15:09:07  LT  java.lang.reflect.InaccessibleObjectException: Unable to make field protected volatile java.io.InputStream java.io.FilterInputStream.in accessible: module java.base does not "opens java.io" to unnamed module @5bfbf16f
```

So, I updated this PR and added a second `--add-options` for hotspot. This made the `InaccessibleObjectException` error go away from the hotspot run now: https://ci.adoptopenjdk.net/view/Test_grinder/job/Grinder/5883/console 

However, there are failures in the Mauve test still, such as : 

```
5:28:47  LT  FAIL: gnu.testlet.java.util.HashMap.AcuniaHashMapTest: no exception wanted !!!, got java.lang.IllegalAccessException: class gnu.testlet.java.util.HashMap.AcuniaHashMapTest cannot access a member of class java.util.HashMap (in module java.base) with modifiers "final" (number 0)
```

Are these unrelated to this PR? 

(Note: tested addition of the second `--ad-options` change on OpenJ9. The MiniMix test is green on OpenJ9 jdk16: Please see Internal Grinder/12993).

pshipton commented:
You can try adding `--illegal-access=permit` to see if it fixes the Hotspot failure, not sure it will. Using `--illegal-access` isn't a long term solution as the option is expected to be removed in a future version of java. 
a) likely this failure shows OpenJ9 is missing something, although we may already have an issue for it @tajila ?
b) if we can't find a command line option workaround for the Hotspot failure, I expect the test needs to be removed as incompatible with jdk16+

It's also concerning that OpenJ9 doesn't show the FilterInputStream.in error, we'll have to investigate this.
```
15:09:07  LT  java.lang.reflect.InaccessibleObjectException: Unable to make field protected volatile java.io.InputStream java.io.FilterInputStream.in accessible: module java.base does not "opens java.io" to unnamed module @5bfbf16f
15:09:07  LT  	at java.base/java.lang.reflect.AccessibleObject.checkCanSetAccessible(AccessibleObject.java:357)
15:09:07  LT  	at java.base/java.lang.reflect.AccessibleObject.checkCanSetAccessible(AccessibleObject.java:297)
15:09:07  LT  	at java.base/java.lang.reflect.Field.checkCanSetAccessible(Field.java:177)
15:09:07  LT  	at java.base/java.lang.reflect.Field.setAccessible(Field.java:171)
15:09:07  LT  	at gnu.testlet.java.lang.reflect.AccessibleObject.accessible.test(accessible.java:71)
```

pshipton commented:
I think the following `getFloat()` causes the `with modifiers "final"` failure on Hotspot, but still need to verify it. I don't see the test doing anything other than `getFloat()`.

```
		Field loadFactor = HashMap.class.getDeclaredField("loadFactor");
		loadFactor.setAccessible(true);
		HashMap hm = new HashMap();
		float lf = loadFactor.getFloat(hm);
		System.out.println("loadFactor: " + lf);
```

pshipton commented:
I can't reproduce the `with modifiers "final"` failure even when running the whole AcuniaHashMapTest in a loop (using a modified test harness). I assume it has something to do with the Hotspot JIT and the entire system test load.

The other interesting thing is that I had to add `--add-opens java.base/java.util=ALL-UNNAMED` in order to run the test, which hasn't been required in the grinders. Maybe there is something wacky about the access checking, like JIT optimized code doesn't do any checks.

BTW, I expect the `--add-opens` argument changes can be delivered so the OpenJ9 testing can pass.

Mesbah-Alam commented:
> BTW, I expect the --add-opens argument changes can be delivered so the OpenJ9 testing can pass.

@llxia - Is there anything else to check before we can deliver this PR?

Mesbah-Alam commented:
> Maybe there is something wacky about the access checking, like JIT optimized code doesn't do any checks.

@pshipton - Should I raise a separate issue for this?

pshipton commented:
> Should I raise a separate issue for this?

Not so far. The code is OpenJDK code (not OpenJ9), and we see the behavior of missed access control checks on Hotspot as well as OpenJ9, I'm not sure there is any OpenJ9 problem.

llxia reviewed (approved):
LGTM
